src/bson.rs

```diff
@@ -486,184 +595,243 @@ impl Bson {
+                            code: code.into(),
+                            scope: scope.clone(),
+                        });
+                    }
+                }
+            }
+
+            ["$timestamp"] => {
+                if let Ok(timestamp) = doc.get_document("$timestamp") {
+                    if let Ok(t) = timestamp.get_i32("t") {
+                        if let Ok(i) = timestamp.get_i32("i") {
+                            return Bson::TimeStamp(TimeStamp {
+                                time: t as u32,
+                                increment: i as u32,
+                            });
+                        }
+                    }
+
+                    if let Ok(t) = timestamp.get_i64("t") {
+                        if let Ok(i) = timestamp.get_i64("i") {
+                            if t >= 0 && i >= 0 {
+                                if t <= (u32::MAX as i64) && i <= (u32::MAX as i64) {
+                                    return Bson::TimeStamp(TimeStamp {
+                                        time: t as u32,
+                                        increment: i as u32,
+                                    });
+                                }
+                            }
+                        }
+                    }
+                }
+            }
+
+            ["$regularExpression"] => {
+                if let Ok(regex) = doc.get_document("$regularExpression") {
+                    if let Ok(pattern) = regex.get_str("pattern") {
+                        if let Ok(options) = regex.get_str("options") {
+                            let mut options: Vec<_> = options.chars().collect();
+                            options.sort();
+
+                            return Bson::Regex(Regex {
+                                pattern: pattern.into(),
+                                options: options.into_iter().collect(),
+                            });
+                        }
+                    }
                 }
-            } else if let Ok(min) = values.get_i64("$minKey") {
-                if min == 1 {
+            }
+
+            ["$dbPointer"] => {
+                if let Ok(db_pointer) = doc.get_document("$dbPointer") {
+                    if let Ok(ns) = db_pointer.get_str("$ref") {
+                        if let Ok(id) = db_pointer.get_object_id("$id") {
+                            return Bson::DbPointer(DbPointer {
+                                namespace: ns.into(),
+                                id: id.clone(),
+                            });
+                        }
+                    }
+                }
+            }
+
+            ["$date"] => {
+                if let Ok(date) = doc.get_i64("$date") {
+                    let mut num_secs = date / 1000;
+                    let mut num_millis = date % 1000;
+
+                    // The chrono API only lets us create a DateTime with an i64 number of
```

Comment:
This might be the largest ratio of size of comment to size of code the comment is documenting that I've eve written.